File: python/examples/demo_worker.py
from __future__ import annotations

import os
import socket
from collections.abc import Mapping
from typing import Any
from uuid import uuid4

import psycopg

from workhorse import HandlerContext, Json, Worker, run_worker_process

LANGUAGE_JOB_TYPE = "demo.language-worker"
SHARED_JOB_TYPE = "demo.shared-worker"
PYTHON_QUEUE = "demo-python"
SHARED_QUEUE = "demo-shared"
SCHEDULE_NAMESPACE = "workhorse-demo"
WORKER_CONCURRENCY = 3
DEFAULT_POLL_MS = 15_000


def database_url(environment: Mapping[str, str] = os.environ) -> str:
    value = environment.get("DATABASE_URL_PRIMARY")
    if not value:
        raise RuntimeError("DATABASE_URL_PRIMARY is required")
    return value


def language_job(payload: Any, context: HandlerContext) -> dict[str, Json]:
    if not isinstance(payload, dict) or payload.get("language") != "python":
        raise ValueError("Python worker received a job for another language")
    return {"language": "python", "runtime": "python", "attempt": context.job.attempt}


def shared_job(payload: Any, context: HandlerContext) -> dict[str, Json]:
    if not isinstance(payload, dict) or not isinstance(payload.get("source"), str):
        raise ValueError("Shared worker requires a source")
    return {"source": payload["source"], "runtime": "python", "attempt": context.job.attempt}


def worker_id() -> str:
    hostname = "".join(
        character if character.isalnum() or character in ".-_" else "-"
        for character in socket.gethostname()
    )
    return f"demo-python-{hostname or 'unknown-host'}-{os.getpid()}-{str(uuid4())[:8]}"


def main() -> None:
    poll_ms = int(os.environ.get("WORKHORSE_WORKER_POLL_MS", DEFAULT_POLL_MS))
    with psycopg.connect(database_url(), autocommit=True) as connection:
        worker = (
            Worker(
                connection,
                queues=(PYTHON_QUEUE, SHARED_QUEUE),
                worker_id=worker_id(),
                concurrency=WORKER_CONCURRENCY,
                poll_ms=poll_ms,
                schedule_namespaces=(SCHEDULE_NAMESPACE,),
                maintenance_interval_ms=1_000,
                registry_interval_ms=250,
            )
            .handle(LANGUAGE_JOB_TYPE, language_job)
            .handle(SHARED_JOB_TYPE, shared_job)
        )
        run_worker_process(worker)


if __name__ == "__main__":
    main()

File: typescript/examples/demo-worker.mjs
import { randomUUID } from "node:crypto";
import { hostname } from "node:os";
import { setTimeout as delay } from "node:timers/promises";
import { pathToFileURL } from "node:url";
import path from "node:path";
import {
  Admin,
  createWorkhorseAdapter,
  defineWorkerProcess,
  Pool,
  Queue,
  runWorkerProcess,
  startWorkerProcess,
} from "@stablemates/workhorse";

const LANGUAGE_TASK_TYPE = "demo.language-worker";
const SHARED_TASK_TYPE = "demo.shared-worker";
const TYPESCRIPT_QUEUE = "demo-typescript";
const SHARED_QUEUE = "demo-shared";

export function createDemoWorker(databaseUrl) {
  const workerId = `demo-typescript-${hostname().replaceAll(/[^\w.-]/g, "-")}-${process.pid}-${randomUUID().slice(0, 8)}`;
  return defineWorkerProcess({
    adapter() {
      const pool = new Pool({ connectionString: databaseUrl, max: 5 });
      return createWorkhorseAdapter({
        database: pool,
        adaptTransaction: (transaction) => transaction,
        close: () => pool.end(),
      });
    },
    workers: [
      {
        options: {
          queues: [TYPESCRIPT_QUEUE, SHARED_QUEUE],
          workerId,
          concurrency: 3,
          pollMs: Number(process.env.WORKHORSE_WORKER_POLL_MS ?? 15_000),
          scheduleNamespaces: ["workhorse-demo"],
        },
        configure(worker) {
          worker
            .handle(LANGUAGE_TASK_TYPE, async (payload, context) => {
              if (payload.language !== "typescript") {
                throw new TypeError("TypeScript worker received a task for another language");
              }
              return { language: "typescript", runtime: "node", attempt: context.task.attempt };
            })
            .handle(SHARED_TASK_TYPE, async (payload, context) => {
              if (typeof payload.source !== "string") {
                throw new TypeError("Shared worker requires a source");
              }
              return { source: payload.source, runtime: "node", attempt: context.task.attempt };
            });
        },
      },
    ],
  });
}

async function verifyDemoWorker(databaseUrl) {
  const runtime = await startWorkerProcess(createDemoWorker(databaseUrl));
  const observer = new Pool({ connectionString: databaseUrl });
  try {
    const queue = new Queue(observer);
    const admin = new Admin(observer);
    const taskIds = await Promise.all([
      queue.enqueue(LANGUAGE_TASK_TYPE, { language: "typescript" }, { queue: TYPESCRIPT_QUEUE }),
      queue.enqueue(SHARED_TASK_TYPE, { source: "typescript-example" }, { queue: SHARED_QUEUE }),
    ]);
    for (let pass = 0; pass < 100; pass += 1) {
      const tasks = await Promise.all(taskIds.map((taskId) => admin.getTask(taskId)));
      if (tasks.every((task) => task?.state === "succeeded")) {
        return { taskIds, results: tasks.map((task) => task.result) };
      }
      const terminalFailure = tasks.find(
        (task) => task?.state === "failed" || task?.state === "canceled",
      );
      if (terminalFailure) throw new Error(`Demo worker task finished in ${terminalFailure.state}`);
      await delay(20);
    }
    throw new Error("Demo worker tasks did not finish");
  } finally {
    await runtime.shutdown();
    await observer.end();
  }
}

const invokedPath = process.argv[1] ? pathToFileURL(path.resolve(process.argv[1])).href : null;
if (import.meta.url === invokedPath) {
  const databaseUrl = process.env.DATABASE_URL;
  if (!databaseUrl) throw new Error("DATABASE_URL is required");
  if (process.argv.includes("--verify")) {
    process.stdout.write(`${JSON.stringify(await verifyDemoWorker(databaseUrl))}\n`);
  } else {
    await runWorkerProcess(createDemoWorker(databaseUrl));
  }
}
